### PR TITLE
Internal Tags

### PR DESCRIPTION
**Pull request by acburdine**
refs #6165

TODO:
- [x] update existing code to use 'visibility' property
- [x] ensure code works
  - [x] client-side setup
  - [x] internal tags don't have `/tag/-slug-/` pages
  - [x] internal tags don't show up in post tags
  - [x] `{{#has}}` helper works with internal tags
- [x] tests

### CODE REVIEW

**Comment by kevinansfield** [comment]
@acburdine looks good, thanks! 😄 Unfortunately I think something has been missed from our conversations (sorry it wasn't documented better!) where there shouldn't be any checkbox in the tag settings menu. Instead any new tag created that begins with a # will be internal so I think that behaviour can be moved out of the PSM and into the Tag model so it's universal. I believe that also means that the visibility attribute will be ignored when passed to the API so that the server will have the final say based on the name & labs flag. (@ErisDS c/d?)

We'll also need a label added to internal tags in the tags screen list because people migrating with old data may have hash tags which are not internal so there needs to be some way to visibly distinguish internal tags.

**Comment by acburdine** [comment]
Holding until after 0.8.0 is released

File: core/server/controllers/frontend/channel-config.js [hunk 1]
```diff
@@ -14,13 +13,13 @@ channelConfig = function channelConfig() {
             name: 'tag',
             route: '/' + config.routeKeywords.tag + '/:slug/',
             postOptions: {
-                filter: labs.isSet('hiddenTags') ? 'tags:\'%s\'+tags.hidden:false' : 'tags:\'%s\''
+                filter: 'tags:\'%s\''
```
Comment (ErisDS) [inline core/server/controllers/frontend/channel-config.js]: I have a slightly different, and I _think_ simpler solution for this on my branch: https://github.com/ErisDS/Ghost/commit/7dff6618fe415c03f94679c8308e8580e7a55334#diff-d23ab2c11f473dd33cad0f93fb92da6cR17

File: core/server/models/base/index.js [hunk 2]
```diff
@@ -459,6 +460,10 @@ ghostBookshelf.Model = ghostBookshelf.Model.extend({
             slug = (slug.indexOf('-') > -1) ? slug.substr(0, slug.indexOf('-')) : slug;
         }
 
+        if (labs.isSet('internalTags') && baseName === 'tag' && /^#/.test(base)) {
```
Comment (ErisDS) [inline core/server/models/base/index.js]: I think this probably warrants a comment 😉

File: core/server/controllers/frontend/channel-config.js (hunk 1 above)
Comment (ErisDS) [inline core/server/controllers/frontend/channel-config.js]: Hmm... I just tested and your comment on the code below is correct - it doesn't get updated properly - that's weird I'm sure I tested that!

Comment (ErisDS) [inline core/server/controllers/frontend/channel-config.js]: This points at a deeper issue, as the config is meant to be dynamic, that's why it's in a function.. but I can deal with that separately of this PR - this solution definitely works.

File: core/test/unit/controllers/frontend/channel-config_spec.js [hunk 3]
```diff
@@ -31,24 +21,4 @@ describe('Channel Config', function () {
         should.exist(result);
         result.name.should.eql('tag');
     });
-
-    describe('hidden tags labs flag', function () {
```
Comment (ErisDS) [inline core/test/unit/controllers/frontend/channel-config_spec.js]: It would be great if we could keep the equivalent of these tests somehow

**Comment by ErisDS** [comment]
Ok - so I think this and pr/4 are pretty much ready to merge - I have left only just minor comments on them both.

What I was working on, when I got utterly stuck, was some improvements to the original spec. I have now written those up [here](https://github.com/TryGhost/Ghost/issues/6165#issuecomment-225272144).

I'm working on rebasing the working changeset I had on top of this PR so I can show you where I got to  with it.

**Comment by ErisDS** [comment]
So... the contents of my stash for this are now [here](https://github.com/ErisDS/Ghost/commit/371b919941b957f10400e9267ab487de0baf90da)

If you're wondering what the fuck.. well you're right. Implementing this for tags was a doddle... but the foreach helper just wouldn't work. 

The problem was something to do with the object containing the resources to loop over that I got back from the foreach helper. It can be an object or an array, but if it's an array, it's one that's been modified with a `length` property. 

I could remove items from the object, but I could not update the length property. I've had an idea though, so I'm gonna have another go just now - and come back to you in a few mins.

**Comment by acburdine** [comment]
@ErisDS fixed all the comments

File: core/server/models/base/index.js [hunk 4]
```diff
@@ -494,6 +495,13 @@ ghostBookshelf.Model = ghostBookshelf.Model.extend({
             slug = (slug.indexOf('-') > -1) ? slug.substr(0, slug.indexOf('-')) : slug;
         }
 
+        // TODO: remove the labs requirement when internal tags is out of beta
+        // This checks if the first character of a tag name is a #. If it is, this
+        // is an internal tag, and as such we should add 'hash' to the beginning of the slug
+        if (labs.isSet('internalTags') && baseName === 'tag' && /^#/.test(base)) {
```
Comment (ErisDS) [inline core/server/models/base/index.js]: Sorry.. I just thought of this. We should only do this if `options.importing` is not set, the same as [here](https://github.com/acburdine/Ghost/blob/ab9d0b646485813faee766296a675f6e933d32d8/core/server/utils/index.js#L72).

File: core/server/models/base/index.js [hunk 5]
```diff
@@ -494,6 +495,15 @@ ghostBookshelf.Model = ghostBookshelf.Model.extend({
             slug = (slug.indexOf('-') > -1) ? slug.substr(0, slug.indexOf('-')) : slug;
         }
 
+        if (!_.has(options, 'importing') || !options.importing) {
```
Comment (acburdine) [inline core/server/models/base/index.js]: @ErisDS is this what you were intending?

Comment (ErisDS) [inline core/server/models/base/index.js]: yes 👍  thanks!

**Pull request merged**